lib/pool.js

```diff
@@ -201,23 +205,60 @@ function getConnection(a1, a2) {
 
 getConnectionPromisified = nodbUtil.promisify(getConnection);
 
-function close(closeCb) {
+function closeWrapper(mode, closeCb) {
   var self = this;
-
-  nodbUtil.assert(arguments.length === 1, 'NJS-009');
-  nodbUtil.assert(typeof closeCb === 'function', 'NJS-006', 1);
-
-  self._close(function(err) {
+  self._close({forceClose:mode}, function (err) {
     if (!err) {
       self._isValid = false;
 
       self.emit('_after_close', self);
     }
-
     closeCb(err);
   });
 }
 
+function close(a1, a2) {
+  var self = this;
+  var drainTime;
+  var closeCb;
+
+  //asserts whether there are 1 or 2 arguments, throws error in case this fails
+  nodbUtil.assert(arguments.length === 1 || arguments.length === 2, 'NJS-009');
+
+  switch (arguments.length) {
+    case 1:
+      nodbUtil.assert(typeof a1 === 'function', 'NJS-006', 1);
+      closeCb = a1;
+      break;
+    case 2:
+      nodbUtil.assert(typeof a1 === 'number', 'NJS-006', 1);
+      nodbUtil.assert(typeof a2 === 'function', 'NJS-006', 2);      
+      closeCb = a2;
+
+      //if connectionsOut is === 0, closePool is called right away, as there is no need to wait at all.
+      drainTime = self._connectionsOut === 0 ? 0 : a1 * 1000;  // a1 represents seconds of time
+
+      // SetTimeout does not accept numbers greater than a 32-bit signed integer.
+      const upperLimitTimeout = 2147483647;
+      nodbUtil.assert(drainTime >= 0 && drainTime <= upperLimitTimeout, 'NJS-005', 1);
+      break;
+  }
+
+  if (arguments.length === 1) {
+    self.closing = false;
```

Comment:
What about moving the setting of the `closing` property to `closeWrapper()`?  It's a straight assignment of the value of `mode`.

Reply:
Or maybe not.  It should probably be set before making an async call